### ISSUE TITLE
SC2319 - changed with Shellcheck 0.10.0

### ISSUE THREAD

**Issue by escoand** [koalaman/shellcheck#2937]
Shellcheck in 0.10.0 complains about an error which is explicit mentioned in the wiki as not relevant. Shellcheck 0.9.0 and also shellcheck.net don't warn.
But this is just happening in some cases, see below.

#### For bugs
- Rule Id (if any, e.g. SC1000): SC2319 
- My shellcheck version (`shellcheck --version` or "online"): 0.10.0
- [x] The rule's wiki page does not already cover this (e.g. https://shellcheck.net/wiki/SC2086)
- [ ] I tried on https://www.shellcheck.net/ and verified that this is still a problem on the latest commit

#### For new checks and feature suggestions
- [ ] https://www.shellcheck.net/ (i.e. the latest commit) currently gives no useful warnings about this
- [ ] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

#### Here's a snippet or screenshot that shows the problem:
![image](https://github.com/koalaman/shellcheck/assets/428567/78134e11-4929-40dd-b165-1d6202f0bc62)

```sh
#!/bin/sh

test -z /test
echo $?
foo=$(echo)
export foo

test -z /test
echo $?
```

#### Here's what shellcheck currently says:
```
In test.sh line 4:
echo $?
     ^-- SC2319 (warning): This $? refers to a condition, not a command. Assign to a variable to avoid it being overwritten.
```

#### Here's what I wanted or expected to see:
nothing

**Comment by cbf123** [comment]
Just ran into an instance of shellcheck 0.10.0 complaining throwing an SC2319 warning about this code:

        [ -f /var/run/.network_upgrade_bootstrap ]
        upgr_bootstrap=$?
        
https://www.shellcheck.net/ thinks it's fine.

**Comment by Stikus** [comment]
Looks like https://github.com/koalaman/shellcheck/issues/2952 is duplicate of this issue.

@koalaman @brother Can anyone take a look? Problem is still here.

**Comment by serge2016** [comment]
Same here! Hope to see the fix... Help, please!

**Comment by bbrk24** [comment]
I also ran into this warning where it was unexpected, but in a different way to the original post:
```sh

**Comment by serge2016** [comment]
Still hope for a fix.

**Comment by FWDekker** [comment]
The issue seems to occur when the assignment and subsequent usage of `$?` are interrupted by a statement that alters `$?`.

For example, the following gives the warning SC2319 on line 2:
```shell
test "$1" != "$2"
foo="$?"
bar="$(cat "bar.txt")"
echo "$foo" "$bar"
```

but the following does not:
```shell
test "$1" != "$2"
foo="$?"
# bar="$(cat "bar.txt")"
echo "$foo" "$bar"
```

even though in both cases, the value of `$?` is not statically predictable (because `$1` and `$2` are unknown) and the output is properly fixed in `$foo`.

**Comment by ao2** [comment]
I can confirm the behavior observed by @FWDekker and the message is really confusing.

@koalaman do you have an opinion on how this should be handled? Can we consider it a false positive in cases like:

```
test "$1" != "$2"
foo="$?"
bar="$(cat "bar.txt")"
echo "$foo" "$bar"
```

Thanks, Antonio

**Comment by Stikus** [comment]
New version of shellcheck was released, the issue is still in place.

Can anyone fix it? Maybe @koalaman ?
